### ISSUE TITLE
Question: How to obtain the spline curve length for a given domain?

### ISSUE THREAD

**Issue by Kevin-Mattheus-Moerman** [jipolanco/BSplineKit.jl#85]
@jipolanco thanks for developing this package. I've only started testing it but it looks great. 

TLDR: Is it possible to easily obtain the length of a spline curves for a given range? So ideally I'd set up an interpolator a bit like `S = interpolate...`, and then ask what is the length of `S` between `x1` and `x2`. Do you have something implemented for that? I suppose we need something like:  

$$L_{a-b} = \int_a^b |f'(t)| dt $$

Below is a longer explanation of what I am trying to use this for. 

I am trying to use interpolation to evenly sample a non-evenly sampled ND input curve. Below I have an example of a first attempt. Here I parameterise the raw curve using the "along curve distance". I then evenly sample the distance from 0 to the final end distance (length) of the curve and feed that to one of your spline based interpolation methods. The below picture shows this works reasonably well. However, with this setup the distances are computed for a linear fit to the curve, i.e. straight line segments connecting the points. It would be better to fit a spline, then get the length of the spline and resample it evenly in this length/distance space. 

Any suggestions? Thanks! 

```julia
using GeometryBasics
using GLMakie
using LinearAlgebra
using BSplineKit

# Define raw curve
t = range(0,2.0*pi,20)
V = [GeometryBasics.Point{3, Float64}(tt,3.0*sin(tt),0.0) for tt ∈ t]

# Evenly sample curve
n = 50; # Number of points for spline 

L = pushfirst!(cumsum(norm.(diff(V))),0.0) # Along curve distance from start
l = range(0.0,maximum(L),n) #Even along curve distance 

S = interpolate(L, V, BSplineOrder(4),Natural()) # Create interpolator

Vi = S.(l) # Evaluate interpolator at even distance increments

# Visualization
fig = Figure(size = (1200,800))

ax1 = Axis3(fig[1, 1],aspect = :data)
hp1 = scatter!(ax1, V,markersize=25,color=:red)
hp2 = lines!(ax1, V,linewidth=3,color=:red)

ax2 = Axis3(fig[1, 2],aspect = :data)
hp1 = scatter!(ax2, Vi,markersize=20,color=:black)
hp2 = lines!(ax2, Vi,linewidth=3,color=:black)
hp2 = lines!(ax2, V,linewidth=2,color=:red)

fig
```
![Screenshot from 2024-02-26 08-16-51](https://github.com/jipolanco/BSplineKit.jl/assets/8392709/f12a989e-417e-4fd6-97a5-c3c52da05906)

**Comment by jipolanco** [comment]
Hi, I completely understand what you're trying to do as I have been dealing with similar issues recently (parametrisation of 3D curves, and how to achieve something close to arc length parametrisation). Thanks for providing a nice minimal example!

In BSplineKit is no built-in way of doing this. To get the total length of the spline $S(t)$ the easiest would be to use numerical integration to estimate the length:

$$L = \int_a^b |S'(t)| \\, \mathrm{d}t$$

where $S'(t)$ is the derivative of the spline with respect to its parameter $t$ (approximately the arc length here). In your example, $a$ and $b$ would be respectively `L[begin]` and `L[end]`, which you can also get via `a = first(knots(S))` and `b = last(knots(S))`. One can use something like [QuadGK.jl](https://github.com/JuliaMath/QuadGK.jl) to perform the numerical integration:

```julia
using QuadGK

Sp = Derivative() * S  # spline derivative

a, b = first(knots(S)), last(knots(S))
L_total, err = quadgk(a, b; rtol = 1e-8) do t
    norm(Sp(t))
end
```

If I run this on your example, I get `L_total = 13.973590503382189` which, as expected, is slightly larger than your initial estimation of the total length using the straight segment approximation (`L[end] = 13.925687625112573`).

If what you want is to reparametrise the curve based on the actual length of each segment, you could also build a new `L` vector obtained from integrating the length segment-by-segment:

```julia
L_new = fill!(similar(L), 0)
for i ∈ eachindex(L_new)[2:end]
    # Compute length of segment [i, i + 1].
    ta = L[i - 1]
    tb = L[i]
    segment_length, err = quadgk(ta, tb; rtol = 1e-8) do t
        norm(Sp(t))  # integrate |S'(t)| in segment [i, i + 1]
    end
    L_new[i] = L_new[i - 1] + segment_length
end

L_new - L  # note: L_new[i] > L[i] for all i since the actual lengths are larger than approximating by straight segments

# We can now interpolate a new spline using the new lengths
S = interpolate(L_new, V, BSplineOrder(4), Natural())
```

If you really care about precision, you could do this iteratively multiple times until `L_new` converges to some constant.

**Comment by jipolanco** [comment]
That sounds interesting! I'd be happy to contribute to your package (and to a possible paper), so let me know when the package is ready to accept contributions :)

**Comment by Kevin-Mattheus-Moerman** [comment]
Question was answered so can be closed.
